README
0. Contents
-----------
1. Compiling
2. Testing
3. Running at boot
4. Mimetype configuration
5. Contact

1. Compiling
------------

Open the file "Makefile" in your favourite text editor and configure the
following options to your tastes.

#Enable this to use libmagic for automatic mimetype detection
LIBMAGIC=no
#Enable this to use zlib for gzip compression
ZLIB=no
#Enable this to use the sendfile() function instead of read()/write()
SENDFILE=no
#Set this to the bin directory you want serve installed in
BINDIR=/usr/bin
#Set this to the directory you want serve_mimetypes installed in
ETCDIR=/etc

After configuring these options run "make".
If all goes according to plan the serve binary should be available in src/serve.

You can now run "make install" as root to install the server binary to
$BINDIR/serve and the mimetypes file to /etc/serve_mimetypes.

2. Testing
----------

After you've compiled and installed serve, you are ready to run it. To test it
out, try:

$ cd ~/public_html
$ serve

and go to http://localhost:8080/ in your browser to check that it's working.

3. Running at boot
------------------

Once you're satisfied that serve itself is working, and if you want it to run
automatically at boot, you will need to write an init script that does something
like this:

 cd $DOC_ROOT
 serve -d -p$PORT -u$USER -g$GROUP -P$PIDFILE -s$HOST >>$LOGFILE 2>>$ERRFILE

The options are as follows:
 -d       Daemonize
 $PORT    Set the TCP port to listen on (e.g. 80)
 $USER    Set the username to setuid to (e.g www-data)
 $GROUP   Set the group name to setgid to (e.g. www-data)
 $PIDFILE Set the path to write a pidfile in (e.g. /var/run/serve.pid)
 $HOST    Set the hostname of the server (sometimes useful for HTTP/1.0 clients)
 $DOC_ROOT Set the location of the root directory (e.g. /var/www)
 $LOGFILE Redirects stdout to the log file (e.g. /var/log/serve.out)
 $ERRFILE Redirects stderr to the error file (e.g. /var/log/serve.err)

See the output of "serve -h" for more options.

The script that kills serve at shutdown should do something like:
 kill `cat $PIDFILE`

4. Mimetype configuration
-------------------------

Even if you compiled serve with libmagic enabled, you will still need a
mimetypes file if you want serve to automatically find index pages. When looking
for an index page, serve goes down the list of file extensions specified in
$DOC_ROOT/.mimetypes, and then /etc/serve_mimetypes (in that order),
trying "index.EXT" where EXT is an extension loaded from those files. The first
existing file that serve can open for reading is used.

$DOC_ROOT is whatever directory serve was started from.

If you want to specify another mimetypes file to serve, in addition to the two
mentioned above, you can give serve the -m option, like so:

$ serve -m ~/my_mimetypes

Serve will use mimetypes in this file in preference to those in the other files.
If you specify several -m options, the later ones are given higher preference
than the earlier ones.

Note that when you change a mimetypes file, you will need to restart serve if
you want it to reload the file.

5. Contact
----------

See the AUTHORS file for information on how to contact me.


ChangeLog
Changelog for serve
-------------------

serve/0.7.4:
 - Now URL decodes properly
 - Switched to a plain Makefile-based build because autotools has too much cruft

serve/0.7.3:
 - Fixes to serve_mimetypes file
 - Directory listing now hides dotfiles, and also closes all the HTML tags

serve/0.7.2: 08-04-2010
 - Gzip compression now works entirely hopefully maybe

serve/0.7.1: 08-04-2010
 - Fixes to gzip to make it work
 - Fixes to build/install environment to make it work for other people

serve/0.7: 08-04-2010
 - Now goes in to the correct directory to run CGI scripts
 - Made add_env() slightly faster
 - Tested NPH scripts for the first time, a few bugs fixed
 - Made a name-value structure for headers, instead of using the stupid
   HEADER() and VALUE() macros
 - Added serve_assert() macro; it's like assert.h's assert(), but it writes
   proper log entries
 - Directory listings with GET query strings now work
 - Now actually does ignore the Last-Modified header for CGI scripts, like it
   was supposed to in 0.6
 - Now includes <netinet/in.h>
 - Only defines a MIN macro if one isn't defined already
 - Only links with -lcrypt on Linux
 - No longer includes <crypt.h>
 - Now includes <sys/param.h>
 - Now includes <limits.h>
 - Now defines _NETBSD_SOURCE to gain access to the dirent d_type macros in
   NetBSD
 - Added a fairly important comment about strdup2() semantics
 - Switched to Makefile-based build
 - No longer bothers gzipping CGI script output if it is less than
   GZIP_BUF_SIZE (i.e. less than 16K)
 - Simplified generation of directory listing paths
 - No longer shows ".." in directory listing of /
 - No longer causes a redirect when links to directories in directory listings
   are followed
 - Introduced "-l" option to specify address to listen on
 - Now uses mlock() to prevent cleartext passwords from being swapped to disk
   (this is stupid for a password that is sent in cleartext over the network!)
 - Now returns "application/octet-stream" for unknown mimetypes when not using
   libmagic instead of the typo "application/octet-strem"
 - Now in public git at http://aw.eso.me.uk/git/ and
   git://aw.eso.me.uk/serve.git
 - Now has a README file
 - Now uses autotools for building, courtesy of Joe Hermaszewski

serve/0.6: 04-08-2009
 - Made the -h description of -s slightly less confusing
 - Can now use sendfile() if USE_SENDFILE is defined at compile-time
 - Ignores the Last-Modified header if there isn't a suitable value. In
   practice this means that CGI scripts no longer have Last-Modified set to
   the current date
 - Wrote bin2c program so that images can be included
 - Now shows the public domain Mosaic/NCSAd/Apache icons in directory listings
 - Removed the request redundancy between mode and meth
 - Lots of additions to the mimetypes file
 - Now returns 400 Bad Request if a POST request is received with no content
   length header
 - Removed disk-backed POST data storage; the kernel can deal with it
 - CGI code simplified a lot
 - No longer segfaults if the client disconnects between specifying the request
   and sending the first header
 - Child processes that received a kill signal no longer attempt to remove the
   pidfile; not that the pidfile ever actually got removed...
 - Now closes the connection after the 400 page if no HTTP version is specified
 - Several bugfixes to do with CGI scripts giving invalid local-server Location
   headers
 - Noticed that Apache doesn't try to handle local redirects itself, serve
   doesn't either any more
 - Now sends 415 Unsupported Media Type if an HTTP POST contains a content
   encoding other than identity
 - Some 500 pages now give a file name, line number, and reason in the error log
 - Now supports gzip content-encoding of CGI scripts and files (but not builtin
   images or error pages) if USE_GZIP is defined at compile-time
 - Now comes with a zlib.supp file for suppressing valgrind messages about
   zlib
 - No longer sends Content-Length responses to HEAD requests if no content
   length is known (e.g. CGI scripts that produce no output on HEAD requests)
 - Fixed concurrency issues in log_text()
 - Removed use of the reserved _t suffix on typedef's
 - Made more code immune to EINTR
 - Now logs all responses with status code < 400 to the log file instead of the
   error file (previously only 200 response codes went to the log file)
 - Now logs the client's user agent and the response content length
 - Now checks that the POST content-length is > 0, instead of != 0

serve/0.5: 17-07-2009
 - Fixed bug whereby output redirection duplicated log entries; identical bug
   (in different place) fixed in 0.2; fully fixed (hopefully) this time
 - Fixed bug whereby children are not killed when the parent is
 - No longer attempts to remove the pidfile if it was unable to be opened for
   writing
 - No longer keeps a pre-existing pidfile open after reading from it
 - No longer continues executing if the pidfile exists and a process with the
   given pid is executing
 - Now returns 400 Bad Request if the requested file does not begin with a /

serve/0.4: 18-06-2009
 - Now accepts any HTTP version with major number 1, rather than just 1.0 and
   1.1; in reality this will make no difference
 - Added more possibilities for sending 500 pages if something goes wrong
 - Made emergency_500() print a message in the error log
 - Made local Location headers from CGI scripts work properly (e.g. Location: /
   should give document root, not filesystem root!)
 - Made authentication necessary (when possible) even when not getting a 200
   response
 - Now sends 400 Bad Request if the client sent unintelligible authentication
   data, rather than 401 Anauthorized
 - Replaced magic length numbers with the decimal_length() macro
 - Made nextline() and log_text() slightly more efficient
 - Added more informative error page texts for the more common error codes
 - Now ignores any whitespace that appears after the "HTTP/1.x" text

src/cgi.c
/* CGI script stuff for serve

   By James Stanley

   Public domain */

#include "serve.h"

/* adds the given name and value to the given environment array
   num must be 1 if there are no items in the array, as the final item must be
   NULL. */
void add_env(char ***env, int num, const char *name, const char *value) {
  int lenname, lenvalue;

  lenname = strlen(name);
  lenvalue = strlen(value);

  /* only resize env if necessary */
  if(num >= INIT_ENV_LENGTH)
    *env = realloc(*env, (num + 1) * sizeof(char*));

  (*env)[num-1] = malloc(lenname + 1 + lenvalue + 1);

  /* name=value */
  strcpy((*env)[num-1], name);
  (*env)[num-1][lenname] = '=';
  strcpy((*env)[num-1] + lenname + 1, value);
  (*env)[num-1][lenname + 1 + lenvalue] = '\0';

  (*env)[num] = NULL;
}

/* free's everything in the given environment array */
void free_env(char **env) {
  int i;

  for(i = 0; env[i]; i++) free(env[i]);
  free(env);
}

/* sets up the environment for the given request */
void setup_env(char ***env, request *r) {
  int num = 1;
  char *scriptname;
  char *pathname;
  char *ptr;
  int i, len = 0, len2 = 0;
  char *header = NULL;
  
  /* make add_env() slightly faster by not needing to realloc every time;
     I profiled this with callgrind and add_env is the second worst performance
     bottleneck when running AjaxTerm */
  *env = malloc(INIT_ENV_LENGTH * sizeof(char*));

  add_env(env, num++, "SERVER_SOFTWARE", SERVER);
  add_env(env, num++, "GATEWAY_INTERFACE", "CGI/1.1");
  add_env(env, num++, "DOCUMENT_ROOT", r->doc_root);
  add_env(env, num++, "SERVER_PORT", port);
  add_env(env, num++, "SERVER_PROTOCOL", r->http);
  add_env(env, num++, "REMOTE_ADDR", r->client);
  add_env(env, num++, "REQUEST_URI", r->reqfile);
  add_env(env, num++, "REQUEST_METHOD", method[r->meth]);

  if(r->auth_realm) {
    add_env(env, num++, "AUTH_TYPE", "Basic");
    add_env(env, num++, "REMOTE_USER", r->auth_user);
  }

  /* get the pathname for SCRIPT_FILENAME */
  if(r->file[0] != '/') {/* make a path name for relative directories */
    len = strlen(r->doc_root);
    len2 = strlen(r->file);
    pathname = malloc(len + 1 + len2 + 1);
    strcpy(pathname, r->doc_root);
    if(pathname[len - 1] != '/') {/* add a slash */
      pathname[len] = '/';
      strcpy(pathname + len + 1, r->file);
    } else {/* no slash necessary */
      strcpy(pathname + len, r->file);
    }
    add_env(env, num++, "SCRIPT_FILENAME", pathname);
    free(pathname);
    len = 0;
    len2 = 0;
  } else {
    add_env(env, num++, "SCRIPT_FILENAME", r->file);
  }

  ptr = strchr(r->reqfile, '?');
  add_env(env, num++, "QUERY_STRING", ptr ? ptr+1 : "");
  if(ptr)	scriptname = strdup2(r->reqfile, ptr - r->reqfile);
  else scriptname = strdup(r->reqfile);
  add_env(env, num++, "SCRIPT_NAME", scriptname);
  free(scriptname);

  /* now the HTTP headers */
  for(i = 0; i < r->num_headers; i++) {
    /* make it long enough */
    len2 = 6 + strlen(r->header_list[i].name);
    if(len2 > len) {
      len = len2;
      header = realloc(header, len);
      strcpy(header, "HTTP_");
    }
    strcpy(header + 5, r->header_list[i].name);
    /* fix case and hyphenation */
    for(ptr = header + 5; *ptr; ptr++) {
      *ptr = toupper(*ptr);
      if(*ptr == '-') *ptr = '_';
    }
    add_env(env, num++, header, r->header_list[i].value);
    /* and check some special ones */
    if(strcmp(header, "HTTP_CONTENT_LENGTH") == 0) {
      add_env(env, num++, "CONTENT_LENGTH", r->header_list[i].value);
    } else if(strcmp(header, "HTTP_CONTENT_TYPE") == 0) {
      add_env(env, num++, "CONTENT_TYPE", r->header_list[i].value);
    } else if(strcmp(header, "HTTP_HOST") == 0) {
      add_env(env, num++, "SERVER_NAME", r->header_list[i].value);
    }
  }
  free(header);
}

/* Runs the CGI script with the given handler */
void run_cgi(request *r) {
  char **env = NULL;
  char buf[1024];
  char *line = NULL;
  char *ptr;
  unsigned long len = 0;
  int n;
  char *handler = r->content_type;
  header *header_list = NULL;
  int num_headers = 0;
  char *sent = NULL;
  int fildes[2];/* 0 is for parent to read from, 1 is for child to write to */
  long long clength = -1;

  /* make a socket to send the post data and cgi output down */
  if(socketpair(AF_UNIX, SOCK_STREAM, 0, fildes) < 0) {
    log_text(err, "Unable to create a socket pair!");
    r->status = 500;
    send_errorpage(r);
    return;
  }

  /* set up the environment */
  setup_env(&env, r);

  /* now fork a handler and give it the file */
  if((n = fork()) < 0) {
    log_text(err, "Unable to fork!");
    r->status = 500;
    send_errorpage(r);
    free_env(env);
    return;
  }

  /* replace child process with script */
  if(!n) exec_script(fildes, env, r);

  /* don't want child's end or the environment */
  free_env(env);
  close(fildes[1]);

  /* write POST data down socket */
  if(r->post_data) write_post_data(fildes[0], r);

  /* sort out headers that don't apply to CGI scripts */
  r->content_length = 0;
  free(r->last_modified);
  r->last_modified = NULL;

  /* read headers until there's a blank line */
  while((n = next_header(fildes[0], &line, &header_list, num_headers++,
                         (size_t *)&clength)) > 0);

  /* sort out header errors */
  if(n == -1) {/* bad headers; non-parsed header script? */
    if(strncmp(line, "HTTP/", 4) != 0) {
      log_text(err, "%s printed a bad header.", r->file);
    }
    send_str(r->fd, line);
    send_str(r->fd, "\r\n");
    free(line);
  } else if(n == -2) {/* premature close of file descriptor */
    log_text(err, "Premature exit by %s while collecting header %d.",
             handler[1] ? handler : r->file, num_headers);
    r->status = 500;
    send_errorpage(r);
    free_headers(header_list, num_headers-1);
    close(fildes[0]);
    return;
  }

  /* next_header didn't add a header if it returned a value < 0 */
  if(n < 0) num_headers--;

  if(n == -1) {/* this is a non-parsed-header script */
    do {
      while((len = read(fildes[0], buf, 1024)) > 0)
        send(r->fd, buf, len, 0);
    } while(len == -1 && errno == EINTR);
    r->close_conn = 1;/* NPH scripts are more reliable if the connection ends */
  } else {
    /* fill in the headers the script gave us */
    sent = malloc(num_headers);

    /* fill_headers returns the content of the Location header */
    if((ptr = fill_headers(r, header_list, num_headers, sent))) {
      /* Location header was sent, redirect */
      if(*ptr == '/') {/* local redirect, handle it ourselves */
	free(r->file);
	free(r->reqfile);
	r->status = 200;
	r->reqfile = ptr;
	r->file = filename(ptr);
	file_stuff(r);
	send_file(r);
      } else {/* not a local redirect, send it to the client */
	r->status = 302;
	r->location = ptr;
	send_errorpage(r);
      }
    } else {/* no location header, send script output */
      r->content_length = 0;

      send_gzipped(r, fildes[0], NOT_MMAPABLE, clength, header_list,
                   num_headers, sent);
    }
  }

  free_headers(header_list, num_headers);
  free(sent);
}

/* writes the post data for the given request to the given file descriptor */
void write_post_data(int fd, request *r) {
  if(!r->post_data) return;

  if(write(fd, r->post_data, r->post_length) < 0) {
    log_text(err, "Unable to write post data to CGI script: %s",
             strerror(errno));
  }
}

/* replaces the current process image with that of the script for the given
   request. */
void exec_script(int *fildes, char * const *env, request *r) {
  char *path;
  char *ptr;
  char *tmp;
  char *handler = r->content_type;

  close(fildes[0]);/* don't want parents end */
  close(r->fd);/* don't give the cgi script our network connection */

  dup2(fildes[1], STDIN_FILENO);
  dup2(fildes[1], STDOUT_FILENO);

  /* now go in to the script's directory */
  tmp = strdup(r->file);
  path = tmp;

  /* strip everything after the right-most slash */
  if((ptr = strrchr(path, '/'))) {
    *ptr = '\0';
    chdir(path);

    /* now get the filename */
    path = ptr + 1;
  }
  
  if(handler[1] == '\0') {/* run script */
    execle(path, path, NULL, env);
    log_text(err, "running script %s with no handler, execle: %s", r->file,
             strerror(errno));
  } else {/* run handler on script */
    execle(handler, handler, path, NULL, env);
    log_text(err, "running handler %s, execle: %s", handler,
             strerror(errno));
  }

  free(tmp);
  close(fildes[1]);
  exit(1);
}

/* sends headers down fd where the corresponding byte in sent is 0.
   There are never enough headers to warrant a bit array for sent */
void send_new_headers(header *header_list, int num_headers, char *sent,
                      int fd) {
  int i;

  for(i = 0; i < num_headers; i++) {
    /* go through headers not sent already */
    if(!sent[i]) {
      send_str(fd, header_list[i].name);
      send_str(fd, ": ");
      send_str(fd, header_list[i].value);
      send_str(fd, "\r\n");
      sent[i] = 1;
    }
  }
}

/* Fills in request structure headers from the header list
   A byte in done is set to 0 if that header does not go in a request structure
   field.
   Returns the content of the Location header, or NULL if there was none */
char *fill_headers(request *r, header *header_list, int num_headers,
                   char *done) {
  int i;
  char *s = NULL;

  memset(done, 1, num_headers);

  for(i = 0; i < num_headers; i++) {
    if(strcasecmp(header_list[i].name, "Content-Type") == 0) {
      free(r->content_type);
      r->content_type = strdup(header_list[i].value);
    } else if(strcasecmp(header_list[i].name, "Content-Length") == 0) {
      r->content_length = strtoull(header_list[i].value, NULL, 10);
    } else if(strcasecmp(header_list[i].name, "Date") == 0) {
      free(r->date);
      r->date = strdup(header_list[i].value);
    } else if(strcasecmp(header_list[i].name, "Last-Modified") == 0) {
      free(r->last_modified);
      r->last_modified = strdup(header_list[i].value);
    } else if(strcasecmp(header_list[i].name, "Status") == 0) {
      r->status = atoi(header_list[i].value);
    } else if(strcasecmp(header_list[i].name, "Location") == 0) {
      free(s);/* get rid of the old one */
      s = strdup(header_list[i].value);
    } else if((strcasecmp(header_list[i].name, "Server") != 0) &&
              (strcasecmp(header_list[i].name, "Connection") != 0)) {
      /* send what the script says instead of letting send_headers do it */
      done[i] = 0;
    }
  }

  return s;
}
